### PR TITLE
Feat/190 implement logic to sortfilter the reinforcement types(2)

### PR DESCRIPTION
**Pull request by ArdtK**
## Issue addressed
Solves #190 

## Code of conduct
- [x] I HAVE NOT added sensitive or compromised (test) data to the repository.
- [x] I HAVE NOT added vulnerabilities to the repository.
- [x] I HAVE discussed my solution with (other) members of the KOSWAT team.

## What has been done?
- Added a new class `StrategyReinforcementInput` that is built when preparing the summary location matrix. A list of `StrategyReinforcementInput` objects is passed to the strategy via the `StrategyInput`.
- Because of this, removed `ground_level_surface` from `StrategyReinforcementCostType`.
- Logic implemented to soft/filter the reinforcements based on the the reinforcements in the `StrategyInput`.

### Checklist
- [x] Tests are either added or updated.
- [x] Branch is up to date with `master`.
- [x] Updated documentation if needed.

## Additional Notes (optional)
Reference data is not impacted.

### CODE REVIEW

File: koswat/strategies/order_strategy/order_strategy.py [hunk 1]
```diff
@@ -46,18 +49,56 @@ def get_default_order_for_reinforcements() -> list[
...
     def get_strategy_order_for_reinforcements(
         self,
+        strategy_reinforcements: list[StrategyReinforcementInput],
     ) -> list[type[ReinforcementProfileProtocol]]:
         """
-        Give the ordered reinforcement types for this strategy,
-        from cheapest to most expensive,
-        possibly omitting reinforcement types that are more expensive and more restrictive than others.
+        Give the ordered reinforcement types for this strategy, from cheapest to most expensive,
+        possibly removing reinforcement types that are more expensive and more restrictive than others.
```
Comment (Carsopre) [inline koswat/strategies/order_strategy/order_strategy.py:56]: Input argument missing in docstring.

File: tests/strategies/order_strategy/test_order_strategy.py [hunk 2]
```diff
@@ -38,15 +40,71 @@ def test_get_default_order_for_reinforcements(self):
...
+        )
+
+        # 3. Verify expectations
+        assert _reinforcements == _expected_result
+        assert _reinforcements[-1] == CofferdamReinforcementProfile
+
+    @pytest.mark.parametrize(
+        "idx", range(len(_default_reinforcements) - 1), ids=_default_reinforcements[:-1]
+    )
+    def test_get_strategy_order_increased_cost_filters_reinforcement(
```
Comment (Carsopre) [inline tests/strategies/order_strategy/test_order_strategy.py:63]: It is not clear to me what are you trying to validate in both tests, please add comments explaining what your expectation is and why should step 2 (run the test data) return the expected value.

File: koswat/strategies/order_strategy/order_strategy.py [hunk 3]
```diff
@@ -46,18 +49,56 @@ def get_default_order_for_reinforcements() -> list[
...
+                if obj and obj.reinforcement_type == CofferdamReinforcementProfile
+            ]
+            _other = [
+                obj
+                for obj in strategy_reinforcements
+                if obj and obj.reinforcement_type != CofferdamReinforcementProfile
+            ]
+            return (_other, _last)
+
+        # Split in a list to be sorted (cheap to expensive) and a list to be put last (Cofferdam for now)
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply koswat/strategies/order_strategy/order_strategy.py:80]: According to requirements the Cofferdam should remain the "last resort" reinforcement profile.

File: koswat/strategies/order_strategy/order_strategy.py [hunk 4]
```diff
@@ -46,18 +49,56 @@ def get_default_order_for_reinforcements() -> list[
...
+        ) -> StrategyReinforcementInput | None:
+            # Only keep the more expensive reinforcement if it is more restrictive
+            if pair[1].ground_level_surface < pair[0].ground_level_surface:
+                return pair[1]
+            return None
+
+        # Check if the current (more expensive) reinforcement is more restrictive than the previous
+        # (the first needs to be prepended as it is always kept)
+        _sorted_pairs = pairwise(_sorted)
+        _kept = [_sorted[0]] + list(map(check_reinforcement, _sorted_pairs))
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply koswat/strategies/order_strategy/order_strategy.py:97]: Added a filter

File: koswat/strategies/order_strategy/order_strategy.py [hunk 5]
```diff
@@ -46,18 +49,56 @@ def get_default_order_for_reinforcements() -> list[
...
+            _last = [
+                obj
+                for obj in strategy_reinforcements
+                if obj and obj.reinforcement_type == CofferdamReinforcementProfile
+            ]
+            _other = [
+                obj
+                for obj in strategy_reinforcements
+                if obj and obj.reinforcement_type != CofferdamReinforcementProfile
+            ]
```
[the remark this answers is not in the document]
Comment (ArdtK) [reply koswat/strategies/order_strategy/order_strategy.py:77]: Looks good

File: tests/strategies/order_strategy/test_order_strategy.py (hunk 2 above)
Comment (ArdtK) [reply tests/strategies/order_strategy/test_order_strategy.py:63]: Added comments

**Review by Carsopre — APPROVED** [review]

**Pull request merged**